### DISCUSSION TITLE
Unable to use <Head> component in my _app for some reason

### DISCUSSION THREAD

**Discussion by ashkenazzio** [vercel/next.js#41712] [Help] [answered]
When trying to use `<Head>` I get: 

`Error: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.`

My `_app` page looks like this:
```
import Head from 'next/head';

function MyApp({ Component, pageProps }) {
  if (Component.getLayout) {
    return Component.getLayout(
      <SessionProvider session={pageProps.session}>
        <ThemeProvider>
          <AlterLayout>
            <Head>
              <link rel='apple-touch-icon' sizes='180x180' />
                // ... etc. head content
            </Head>
            <Component {...pageProps} />
          </AlterLayout>
        </ThemeProvider>
      </SessionProvider>
    );
  }

  return (
      <SessionProvider session={pageProps.session}>
        <ThemeProvider>
            <Layout>
              <Head>
              <link rel='apple-touch-icon' sizes='180x180' />
                // ... etc. head content
              </Head>
              <Component {...pageProps} />
            </Layout>
        </ThemeProvider>
      </SessionProvider>
  );
}
```

**The thing is, opening up pages that use `<AlterLayout>` works fine and `<head>` is added to page source as expected, but opening up pages that use `<Layout>` give me the error.**

I'll also point out that I use `<Head>` to add `<title>` to different pages but removing them doesn't affect the problem.

**Comment by icyJoseph** [comment, marked as answer]
Could you provide a link to a repository? Or a [Stackblitz](https://stackblitz.com/) reproduction?

Things to double check:

- Pages have default exports
- Layout, AlterLayout, etc are correctly imported via named or default exports

## Edit

OP was passing the children prop of Layout to `React.cloneElement`, assuming to have a single a child. Adding any other sibling to `Component` breaks the call.

**Reply by ashkenazzio** [reply]
Sure, here's the project's repository:
_**--removed--**_

- All pages are exported with `export default`
- Layout components are also exported and imported with as  default modules

**Reply by icyJoseph** [reply]
Alright you've fallen trap to `cloneElement`

When you add `Head` you have:

```jsx
       <Layout>
                <Head>
                  <title>Hello World</title>
                </Head>
                <Component {...pageProps} />
       </Layout>
```

Which means that the children is now a list of two elements, and passing a list of elements to `cloneElement` causes errors.

Please refer to the new docs: https://beta.reactjs.org/apis/react/cloneElement for an in depth analysis, and for a formal API review the old docs are a bit better, I think, https://reactjs.org/docs/react-api.html#cloneelement

In particular this bit:

```md
React.cloneElement() is almost equivalent to:

    <element.type {...element.props} {...props}>{children}</element.type>
```

The list of children made by Head and Component has no type property, think of it as a `[]`, no type exists.

This would also crash:

```jsx
            <Layout>
              <div>
                <p>Hello World</p>
              </div>
              <Component {...pageProps} />
            </Layout>
```

Etc, etc...

So you might want to use `Children.map` or `Children.toArray`, or some other method to inject the props you want to, into `Component`.

You might also want to forbid multiple children, https://reactjs.org/docs/react-api.html#reactchildrenonly, etc.

Wrapping the `Layout` children with a fragment, will silently introduce a bug, I think, where the props never pass to `Component`.

**Reply by ashkenazzio** [reply]
Thank you for the elaborated reply 🙏🏻

I chose to go for `Children.map` by replacing this:
```
    {cloneElement(props.children, {
            layout: [setTitle, setFilterElement],
            filter: filter,
          })}
```
with this:
```
   {Children.map(props.children, (child) => {
            return cloneElement(child, {
              layout: [setTitle, setFilterElement],
              filter: filter,
            });
          })}
```
